coordinator/internal/authority/credentials.go

```diff
@@ -0,0 +1,135 @@
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"google.golang.org/grpc/credentials"
+	"k8s.io/utils/clock"
+)
+
+// Credentials are gRPC transport credentials that dynamically update with the Coordinator state.
+type Credentials struct {
+	issuer   atls.Issuer
+	getState func() (*State, error)
+
+	logger                     *slog.Logger
+	attestationFailuresCounter prometheus.Counter
+	kdsGetter                  *snp.CachedHTTPSGetter
+}
+
+// Credentials creates new transport credentials that validate peers according to the latest manifest.
+func (a *Authority) Credentials(reg *prometheus.Registry, issuer atls.Issuer) (*Credentials, func()) {
+	ticker := clock.RealClock{}.NewTicker(24 * time.Hour)
+	kdsGetter := snp.NewCachedHTTPSGetter(memstore.New[string, []byte](), ticker, logger.NewNamed(a.logger, "kds-getter"))
+	attestationFailuresCounter := promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Subsystem: "contrast_meshapi",
+		Name:      "attestation_failures_total",
+		Help:      "Number of attestation failures from workloads to the Coordinator.",
+	})
+
+	return &Credentials{
+		issuer: issuer,
+		getState: func() (*State, error) {
+			if err := a.syncState(); err != nil {
+				return nil, fmt.Errorf("syncing state: %w", err)
+			}
+			state := a.state.Load()
+			if state == nil {
+				return nil, errors.New("coordinator is not initialized")
+			}
+			return state, nil
+		},
+		logger:                     a.logger,
+		attestationFailuresCounter: attestationFailuresCounter,
+		kdsGetter:                  kdsGetter,
+	}, ticker.Stop
+}
+
+// ServerHandshake implements an aTLS handshake for the latest state.
+//
+// If successful, the state will be passed to gRPC as AuthInfo.
+func (c *Credentials) ServerHandshake(rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	state, err := c.getState()
+	if err != nil {
+		return nil, nil, fmt.Errorf("getting state: %w", err)
+	}
+
+	authInfo := AuthInfo{State: state}
+
+	validator := snp.NewValidatorWithCallbacks(state, c.kdsGetter,
+		logger.NewWithAttrs(logger.NewNamed(c.logger, "validator"), map[string]string{"tee-type": "snp"}),
+		c.attestationFailuresCounter, &authInfo)
+
+	serverCfg, err := atls.CreateAttestationServerTLSConfig(c.issuer, []atls.Validator{validator})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	conn, info, err := credentials.NewTLS(serverCfg).ServerHandshake(rawConn)
+	if err != nil {
+		return nil, nil, err
+	}
+	tlsInfo, ok := info.(credentials.TLSInfo)
+	if ok {
+		authInfo.TLSInfo = tlsInfo
+	} else {
+		c.logger.Error("credentials.NewTLS returned unexpected AuthInfo", "obj", info)
+	}
+
+	return conn, authInfo, nil
+}
+
+// Info provides information about the protocol.
+func (c *Credentials) Info() credentials.ProtocolInfo {
+	return credentials.NewTLS(nil).Info()
+}
+
+// Clone the credentials object.
+func (c *Credentials) Clone() credentials.TransportCredentials {
+	cloned := *c
+	return &cloned
+}
```

Comment:
Looking at the [upstream code](https://github.com/grpc/grpc-go/blob/e6b6318ad90847fa8b5ed987b964f1494179c89f/credentials/alts/alts.go#L270), I'm unsure whether a shallow copy is sufficeint here.

Reply:
I agree, and it seems like a bad idea to clone the KDS getter due to the embedded ticker. I checked the gRPC source and found that credentials are only cloned when client-side load balancing, so I decided to remove the implementation.